### PR TITLE
Automatic tslint.json generation with --init command

### PR DESCRIPTION
**Pull request by JKillian**
Implements #717 

A simple implementation of a feature to generate a tslint.json file for users. This PR is very limited in scope - there are lots of ways for this to be improved in the future.

This feature intentionally isn't added to the README yet, that can be done once a better default configuration is settled on.

### CODE REVIEW

**Comment by adidahiya** [comment]
- [ ] needs test in `check-bin.sh`
- [ ] we should settle on a good default config and add it to the readme before merging this

**Comment by JKillian** [comment]
Added test. As far as a default config, I think it should be pretty conservative in what is enabled

**Comment by JKillian** [comment]
Added a commit with a proposed reasonable default config. I looked at our settings as well as the rules TypeScript used as a baseline, and ruling out rules that are more aggressive (or bug prone, like no-unused-variable).

File: src/configuration.ts [hunk 1]
```diff
@@ -19,20 +19,35 @@ import * as fs from "fs";
...
         "no-eval": true,
+        "no-internal-module": true,
         "no-trailing-whitespace": true,
-        "no-unreachable": true,
-        "no-use-before-declare": true,
-        "quotemark": [true, "double"],
-        "semicolon": true
+        "no-var-keyword": true,
+        "one-line": [true, "check-open-brace", "check-whitespace"],
+        "quotemark": [true, "single"],
```
Comment (adidahiya) [inline src/configuration.ts]: why single? we use double and so does the TS team. let's change it back.

File: src/configuration.ts [hunk 2]
```diff
@@ -19,20 +19,35 @@ import * as fs from "fs";
...
+        "semicolon": true,
+        "typedef-whitespace": [true, {
+            "call-signature": "nospace",
+            "index-signature": "nospace",
+            "parameter": "nospace",
+            "property-declaration": "nospace",
+            "variable-declaration": "nospace"
+        }],
```
Comment (adidahiya) [inline src/configuration.ts]: let's add `triple-equals`; it's pretty uncontroversial

Comment (adidahiya) [inline src/configuration.ts]: can you add `typedef-whitespace` to our own tslint.json as well?

File: src/tslint-cli.ts [hunk 3]
```diff
@@ -42,6 +42,10 @@ let processed = optimist
             alias: "help",
             describe: "display detailed help"
         },
+        "i": {
+            alias: "init",
+            describe: "generate a tslint.json config file in the cwd"
```
Comment (adidahiya) [inline src/tslint-cli.ts]: I would expand cwd to 'current working directory'

File: src/tslint-cli.ts [hunk 4]
```diff
@@ -81,6 +85,17 @@ if (argv.v != null) {
     process.exit(0);
 }
 
+if (argv.i != null) {
+    if (fs.existsSync(CONFIG_FILENAME)) {
+        console.error(`Cannot generate ${CONFIG_FILENAME}: file already exists`);
+        process.exit(1);
+    }
+
+    const tslintJSON = JSON.stringify(DEFAULT_CONFIG, undefined, "   ");
```
Comment (adidahiya) [inline src/tslint-cli.ts]: this is 3 spaces; it should be 2 or 4

Comment (adidahiya) [inline src/tslint-cli.ts]: rename to `tslintJson`

File: test/check-bin.sh [hunk 5]
```diff
@@ -43,8 +42,27 @@ expectOut $? 0 "tslint with valid arguments did not exit correctly"
...
+./tslint --init
+if [ ! -f tslint.json ]; then
+  echo "--init failed, tslint.json not created"
+  num_failures=$(expr $num_failures + 1)
+fi
+expectOut $? 0 "tslint with --init flag did not exit correctly"
+
+# should fail since tslint.json already exists
+./tslint --init
+expectOut $? 1 "tslint with --init flag did not exit correctly when tslint.json already exists"
```
Comment (adidahiya) [inline test/check-bin.sh]: :+1:

File: src/configuration.ts (hunk 1 above)
Comment (JKillian) [inline src/configuration.ts]: Single quotes are scientifically proven to be easier for humans on standard english keyboards to type!

File: src/configuration.ts (hunk 2 above)
Comment (JKillian) [inline src/configuration.ts]: :+1:  with the `allow-null-check` option I think

File: src/tslint-cli.ts (hunk 4 above)
Comment (JKillian) [inline src/tslint-cli.ts]: (JS stdlib treats it as JSON, not Json, more consistent to keep it capitalized)

File: src/configuration.ts (hunk 1 above)
Comment (adidahiya) [inline src/configuration.ts]: meh, but it doesn't make sense to have the default config be different from our own code like this. we should discuss moving to single quotes elsewhere... we choose double to make it easier to switch back-and-forth between languages like Java to TS. many others, including the compiler team, use double quotes for the same reason.

File: src/configuration.ts (hunk 2 above)
Comment (adidahiya) [inline src/configuration.ts]: yep

File: src/tslint-cli.ts (hunk 4 above)
Comment (adidahiya) [inline src/tslint-cli.ts]: > JS stdlib treats it as JSON

what does that mean?

The "JSON" substring feels a bit too direct of a conflict with the builtin `JSON` value.

**Comment by JKillian** [comment]
[Changes made]

**Comment by adidahiya** [comment]
:+1:

**Pull request merged**